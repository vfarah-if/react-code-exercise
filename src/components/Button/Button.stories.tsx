import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Button, ButtonProps } from './Button';

export default {
  title: 'Component/Button',
  component: Button,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as Meta;

const Template: Story<ButtonProps> = args => <Button {...args} />;

export const PrimaryButton = Template.bind({});
PrimaryButton.args = {
  primary: true,
  label: 'Button',
};

export const SecondaryButton = Template.bind({});
SecondaryButton.args = {
  label: 'Button',
};

export const LargeButton = Template.bind({});
LargeButton.args = {
  size: 'large',
  label: 'Button',
};

export const SmallButton = Template.bind({});
SmallButton.args = {
  size: 'small',
  label: 'Button',
};

export const LabellessButton = Template.bind({});
LabellessButton.args = {};

export const GreenBackgroundButton = Template.bind({});
GreenBackgroundButton.args = {
  primary: true,
  size: 'small',
  label: 'Custom background',
  backgroundColor: 'green',
};
