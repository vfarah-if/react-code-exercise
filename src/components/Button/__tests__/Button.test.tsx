import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import {
  PrimaryButton,
  SecondaryButton,
  LargeButton,
  SmallButton,
  LabellessButton,
  GreenBackgroundButton,
} from '../Button.stories';

import Button from '../Button';

describe('Button', () => {
  test('should fire a button click event', () => {
    const mockCallBack = jest.fn();
    render(<Button label='Test Click' onClick={mockCallBack} />);

    fireEvent.click(getButton());

    expect(mockCallBack).toHaveBeenCalledTimes(1);
  });

  /* NOTE: These tests could be seen as bad practises because they are not testing behaviour and they
   **      are also using snapshots, which many find to be a bad practise. I like this snapshot and
   **      testing component because it helps to protect the storybook component, story book test drive
   **      the scenarios I created and so is an extension of visual testing as well as unit testing
   */

  test('should create a primary button', () => {
    render(
      <PrimaryButton
        primary={PrimaryButton.args?.primary}
        label={PrimaryButton.args?.label || ''}
      />
    );

    expect(getButton()).toHaveTextContent('Button');
    expect(getButton()).toMatchSnapshot('primary button');
  });

  test('should create a small secondry button', () => {
    render(
      <SmallButton
        size={SmallButton.args?.size}
        label={SmallButton.args?.label || ''}
      />
    );

    expect(getButton()).toMatchSnapshot('small button');
  });

  test('should create a secondry button', () => {
    render(<SecondaryButton label={SmallButton.args?.label || ''} />);

    expect(getButton()).toMatchSnapshot('secondry button');
  });

  test('should create a large button', () => {
    render(
      <LargeButton
        size={LargeButton.args?.size}
        label={LargeButton.args?.label || ''}
      />
    );

    expect(getButton()).toMatchSnapshot('large button');
  });

  test('should create a geen background button', () => {
    render(
      <GreenBackgroundButton
        primary={GreenBackgroundButton.args?.primary || true}
        size={GreenBackgroundButton.args?.size}
        label={GreenBackgroundButton.args?.label || ''}
        backgroundColor={GreenBackgroundButton.args?.backgroundColor}
      />
    );

    expect(getButton()).toMatchSnapshot('green background button');
  });

  test('should create a yellow labeless button as alternative test if devaiting from convenrional tests', () => {
    render(
      <LabellessButton
        label={LabellessButton.args?.label || ''}
        style={{ backgroundColor: 'yellow' }}
      />
    );

    expect(getButton()).toMatchSnapshot('labelless yellow button');
  });
});

function getButton(): Element | Node | Document | Window {
  return screen.getByRole('button');
}
