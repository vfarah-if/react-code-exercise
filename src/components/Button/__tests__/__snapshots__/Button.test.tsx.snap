// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should create a geen background button: green background button 1`] = `
<button
  class="sc-bczRLJ kdthhz"
  label="Custom background"
>
  Custom background
</button>
`;

exports[`Button should create a large button: large button 1`] = `
<button
  class="sc-bczRLJ cGAcOK"
  label="Button"
>
  Button
</button>
`;

exports[`Button should create a primary button: primary button 1`] = `
<button
  class="sc-bczRLJ hKYNbp"
  label="Button"
>
  Button
</button>
`;

exports[`Button should create a secondry button: secondry button 1`] = `
<button
  class="sc-bczRLJ jGJCSX"
  label="Button"
>
  Button
</button>
`;

exports[`Button should create a small secondry button: small button 1`] = `
<button
  class="sc-bczRLJ jcxdVN"
  label="Button"
>
  Button
</button>
`;

exports[`Button should create a yellow labeless button as alternative test if devaiting from convenrional tests: labelless yellow button 1`] = `
<button
  class="sc-bczRLJ jGJCSX"
  label=""
  style="background-color: yellow;"
/>
`;
